internal/providers/instantiator.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"fmt"
+
+	serverconfig "github.com/stacklok/minder/internal/config/server"
+	"github.com/stacklok/minder/internal/crypto"
+	"github.com/stacklok/minder/internal/db"
+	gitclient "github.com/stacklok/minder/internal/providers/git"
+	githubapp "github.com/stacklok/minder/internal/providers/github/app"
+	ghclient "github.com/stacklok/minder/internal/providers/github/oauth"
+	httpclient "github.com/stacklok/minder/internal/providers/http"
+	"github.com/stacklok/minder/internal/providers/ratecache"
+	"github.com/stacklok/minder/internal/providers/telemetry"
+	provinfv1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+// TraitInstantiator is responsible for creating instances of the Profile
+// interfaces defined in `github.com/stacklok/minder/pkg/providers/v1` from
+// the providers stored in the database.
+type TraitInstantiator interface {
+	GetGit(ctx context.Context, provider *db.Provider) (provinfv1.Git, error)
+	GetHTTP(ctx context.Context, provider *db.Provider) (provinfv1.REST, error)
+	GetGitHub(ctx context.Context, provider *db.Provider, options *GitHubOptions) (provinfv1.GitHub, error)
+	GetRepoLister(ctx context.Context, provider *db.Provider) (provinfv1.RepoLister, error)
+}
+
+// GitHubOptions contains options used when creating a GitHub instance
+// If omitted, these will be looked up in the database
+type GitHubOptions struct {
+	Credential  provinfv1.GitHubCredential
+	OwnerFilter *string
+}
```

Comment:
I'm not a fan of having these on `GetGitHub`, rather than having them be part of the TraitInstantiator object state.  Do you have a use case beyond `verifyProviderTokenIdentity`?

Reply:
Yeah, the more I think about the approach I took, the less I like it. I might try another approach to avoid this hack.

Reply:
Looking more at the code, I wonder if `verifyProviderTokenIdentity` should really be `verifyGitHubTokenIdentity`, and it should simply instantiate a `github.Client` itself.

In both cases, that call is being made before the provider is created, so it's a bit of an odd case that might be be addressed by not trying to re-use the rest of the framework.